app/usecases/score.py

```diff
@@ -200,3 +189,100 @@ async def build_full_replay(score: Score) -> Optional[BinaryWriter]:
         .write_raw(raw_data)
         .write_i64_le(score.id)
     )
+
+
+class FirstPlaceRow(NamedTuple):
+    beatmap_md5: str
+    mode: Mode
+    score_id: int
+    user_id: int
+
+
+async def get_user_first_places(user_id: int) -> list[FirstPlaceRow]:
+    first_db = await app.state.services.read_database.fetch_all(
+        "SELECT beatmap_md5, mode, rx, scoreid, userid FROM scores_first "
+        "WHERE userid = :userid",
+        {
+            "userid": user_id,
+        },
+    )
+
+    return [
+        FirstPlaceRow(
+            beatmap_md5=row["beatmap_md5"],
+            mode=Mode.from_rx_mode(row["mode"], row["rx"]),
+            score_id=row["scoreid"],
+            user_id=row["userid"],
+        )
+        for row in first_db
+    ]
+
+
+async def delete_first_place(score_id: int) -> None:
+    await app.state.services.write_database.execute(
+        "DELETE FROM scores_first WHERE scoreid = :score_id",
+        {
+            "score_id": score_id,
+        },
+    )
+
+
+async def insert_first_place(score: Score) -> None:
+    await app.state.services.write_database.execute(
+        (
+            "INSERT INTO scores_first (beatmap_md5, mode, rx, scoreid, userid) VALUES "
+            "(:md5, :mode, :rx, :sid, :uid)"
+        ),
+        {
+            "md5": score.map_md5,
+            "mode": score.mode.as_vn,
+            "rx": score.mode.relax_int,
+            "sid": score.id,
+            "uid": score.user_id,
+        },
+    )
+
+
+async def calculate_first_place(beatmap_md5: str, mode: Mode) -> None:
+    cur_first = await app.state.services.read_database.fetch_one(
```

Comment:
`cur`?

Reply:
Renamed to `current_first`, should be clearer